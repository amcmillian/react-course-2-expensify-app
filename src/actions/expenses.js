import { database } from 'firebase';
import uuid from 'uuid';
import db from '../firebase/firebase';

/*
Without DB...
Component calls Action Generator
Action Generator returns Object
Component dispatches Object
Redux Store changes

With DB...
Component calls Action Generator
Action Generator returns Function
Component dispatches Function
Function runs (has the ability to dispatch other Actions)
*/


// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdOn = 0
        } = expenseData;
        const expense = {description, note, amount, createdOn};
        return db.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

// REMOVE_EXPENSE
export const removeExpense = ({id}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ({id}) => {
    console.log(id);
    return (dispatch) => {
        return db.ref(`expenses/${id}`).remove().then(() => {
            dispatch(removeExpense({id}));
        });
    }
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch) => {
        return db.ref(`expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id, updates));
        });
    };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        return db.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpenses(expenses));
        });
    };
};