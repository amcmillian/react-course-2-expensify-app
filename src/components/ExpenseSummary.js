import React from 'react';
import {connect} from 'react-redux';
import selectExpensesTotal from '../selectors/expenses-total.js';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';

const ExpenseSummary = (props) => {
    const expenseWord = props.expenses.length === 1 ? 'expense' : 'expenses';
    const formattedExpenses = numeral(selectExpensesTotal(props.expenses)/100).format('$0,0.00');
    return (
        <div>
            <h3>
                Viewing {props.expenses.length} {expenseWord} totalling {formattedExpenses}
            </h3>    
        </div>
    )
};

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseSummary);