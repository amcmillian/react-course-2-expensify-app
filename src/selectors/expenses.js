import moment from 'moment';

// Set visible expenses

export default (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdOn);
        // const startDateMatch = (typeof startDate !== 'number') || (expense.createdOn >= startDate);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        // const endDateMatch = (typeof endDate !== 'number') || (expense.createdOn <= endDate);
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = (typeof text !== 'string') || (expense.description.toLowerCase().includes(text.toLowerCase()));
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy == 'date') {
            return a.createdOn < b.createdOn ? 1 : -1;
        } else if (sortBy == 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    })
}