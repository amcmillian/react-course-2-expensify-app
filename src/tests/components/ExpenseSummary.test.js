import React from 'react';
import {shallow} from 'enzyme';
import ExpenseSummary from '../../components/ExpenseSummary';
import expenses from '../fixtures/expenses';

test('should match expense summary snapshot for fixture expenses', () => {
    const wrapper = <ExpenseSummary expenses={expenses} />;
    expect(wrapper).toMatchSnapshot();
});

test('should match expense summary snapshot for 0 expenses', () => {
    const wrapper = <ExpenseSummary expenses={[]} />;
    expect(wrapper).toMatchSnapshot();
});

test('should match expense summary snapshot for 1 expense', () => {
    const wrapper = <ExpenseSummary expenses={expenses[0]} />;
    expect(wrapper).toMatchSnapshot();
});