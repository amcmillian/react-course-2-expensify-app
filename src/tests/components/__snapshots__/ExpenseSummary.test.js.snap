// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match expense summary snapshot for 0 expenses 1`] = `
<Connect(ExpenseSummary)
  expenses={Array []}
/>
`;

exports[`should match expense summary snapshot for 1 expense 1`] = `
<Connect(ExpenseSummary)
  expenses={
    Object {
      "amount": 195,
      "createdOn": 0,
      "description": "Gum",
      "id": "1",
      "note": "",
    }
  }
/>
`;

exports[`should match expense summary snapshot for fixture expenses 1`] = `
<Connect(ExpenseSummary)
  expenses={
    Array [
      Object {
        "amount": 195,
        "createdOn": 0,
        "description": "Gum",
        "id": "1",
        "note": "",
      },
      Object {
        "amount": 109500,
        "createdOn": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "",
      },
      Object {
        "amount": 4500,
        "createdOn": 345600000,
        "description": "Credit Card",
        "id": "3",
        "note": "",
      },
    ]
  }
/>
`;
